src/canard.c

```diff
@@ -1,7 +1,919 @@
+/*
+ * Copyright (c) 2016 UAVCAN Team
+ *
+ * Distributed under the MIT License, available in the file LICENSE.
+ *
+ * Author: Michael Sierra <sierramichael.a@gmail.com>
+ *
+ */
+
 #include "canard.h"
 #include "canard_internals.h"
+#include <inttypes.h>
+
+
+#define CANARD_MAKE_TRANSFER_DESCRIPTOR(data_type_id, transfer_type, \
+                                        src_node_id, dst_node_id) \
+    ((data_type_id) | ((transfer_type) << 16) | \
+     ((src_node_id) << 18) | ((dst_node_id) << 25))
+
+struct CanardTxQueueItem
+{
+    CanardTxQueueItem* next;
+    CanardCANFrame frame;
+};
+
+/**
+ *  API functions
+ */
+
+/**
+ * Initializes the library state.
+ * Local node ID will be set to zero, i.e. the node will be anonymous.
+ */
+void canardInit(CanardInstance* out_ins,  void* mem_arena, size_t mem_arena_size,
+                CanardOnTransferReception on_reception, CanardShouldAcceptTransfer should_accept)
+{
+    out_ins->node_id = CANARD_BROADCAST_NODE_ID;
+    out_ins->on_reception = on_reception;
+    out_ins->should_accept = should_accept;
+    out_ins->rx_states = NULL;
+    out_ins->tx_queue = NULL;
+    canardInitPoolAllocator(&out_ins->allocator, mem_arena, mem_arena_size / CANARD_MEM_BLOCK_SIZE);
+}
+
+/**
+ * Assigns a new node ID value to the current node.
+ */
+void canardSetLocalNodeID(CanardInstance* ins, uint8_t self_node_id)
+{
+    ins->node_id = self_node_id;
+}
+
+/**
+ * Returns node ID of the local node.
+ * Returns zero if the node ID has not been set.
+ */
+uint8_t canardGetLocalNodeID(const CanardInstance* ins)
+{
+    return ins->node_id;
+}
+
+/**
+ * Sends a broadcast transfer.
+ * If the node is in passive mode, only single frame transfers will be allowed.
+ */
+int canardBroadcast(CanardInstance* ins,
+                    uint64_t data_type_signature,
+                    uint16_t data_type_id,
+                    uint8_t* inout_transfer_id,
+                    uint8_t priority,
+                    const void* payload,
+                    uint16_t payload_len)
+{
+    uint32_t can_id;
+    uint16_t crc = 0xFFFFU;
+
+    if (payload == NULL)
+    {
+        return -1;
+    }
+    if (priority > 31)
+    {
+        return -1;
+    }
+    if (canardGetLocalNodeID(ins) == 0)
+    {
+        if (payload_len > 7)
+        {
+            return -1;
+        }
+        else
+        {
+            // anonymous transfer, random discriminator
+            uint16_t discriminator = (crcAdd(0xFFFFU, payload, payload_len)) & 0x7FFEU;
+            can_id = ((uint32_t)priority << 24) | (uint32_t)(discriminator << 9) |
```

Comment:
`(uint32_t)(discriminator << 9)` - will fail on 16/8-bit systems because of type range limits. Correct: `((uint32_t)discriminator << 9)`